
# 作業目標:
    
    了解keras 模型架構


# 作業重點:
    
    請修改input shape: (Conv2D(64, (3, 3))的設定, 新增一層 Dense 並觀看 model.summary 的輸出


In [1]:
# 載入必須使用的 Library
import keras
from keras.models import Sequential
from keras.datasets import cifar10
from keras.layers import Dense, Dropout, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D

Using TensorFlow backend.


In [2]:
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(50000, 32, 32, 3)
(50000, 1)
(10000, 32, 32, 3)
(10000, 1)


In [3]:
y_train = keras.utils.to_categorical(y_train)
y_test = keras.utils.to_categorical(y_test)
print(y_train.shape)
print(y_test.shape)

(50000, 10)
(10000, 10)


In [4]:
num_class = 10

In [5]:
# build our CNN model, 多加幾層
model  = Sequential()
model.add(Conv2D(128, (2, 2), 
                 input_shape=x_train.shape[1:]))
model.add(Activation('relu'))

model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
#model.add(Dropout(0.5))

model.add(Dense(256))
model.add(Activation('relu'))
#model.add(Dropout(0.2))

model.add(Dense(num_class))
model.add(Activation('softmax'))


W1118 19:55:13.898906 13104 deprecation_wrapper.py:119] From C:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W1118 19:55:13.934885 13104 deprecation_wrapper.py:119] From C:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W1118 19:55:13.937883 13104 deprecation_wrapper.py:119] From C:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.



In [6]:
print(model.summary())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 31, 31, 128)       1664      
_________________________________________________________________
activation_1 (Activation)    (None, 31, 31, 128)       0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 123008)            0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               62980608  
_________________________________________________________________
activation_2 (Activation)    (None, 512)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 256)               131328    
_________________________________________________________________
activation_3 (Activation)    (None, 256)               0         
__________

### Caculate Parameter
- 前言:這裡指的Parameter是weight和bias。在不同層的狀態下會產生不同的Parameter值
***
- Input Layer:輸入層並沒有進行學習，因此不會產生weights和bias，也就不會有Parameter
- Convolutional Layer:卷積層中會有filter來對輸入值進行卷積(convolution)，我們通常定義:
    1. filter size =$ n * m $(但通常filter size長寬會是相等) = $n^2$
    2. feature maps = $k$(這裡指的是filter的個數，相當於深度；同樣的可以用在Input data的深度，像是RGB)
    3. feature maps = $l$(這裡指的是Input data的深度，RGB)
    4. Parameters:經過Convolutional Layer產生的weights和bias總和:
        $$weights=(n * m * l) * k$$
        $$bias = k$$
        $$Parameter = weights + bias = (n * m * l + 1) * k$$
        
    5. Example: input size = (32, 32, 3), filter size = (2, 2), filter number = 128, we know $(l=3)$, $(n,m=2)$,  $(k=128)$  
        So parameter = $(2 * 2 * 3 + 1)* 128 = 13*128 = 1664$
***        
- Referance:
    1. [How to calculate the number of parameters in the CNN](https://medium.com/@iamvarman/how-to-calculate-the-number-of-parameters-in-the-cnn-5bd55364d7ca)